src/mutation.cpp

```diff
@@ -463,17 +733,114 @@ void player::activate_mutation( std::string mut )
             newit = i_add(newit);
             add_msg(m_info, "%c - %s", newit.invlet == 0 ? ' ' : newit.invlet, newit.tname().c_str());
         }
-        traits[mut].powered = false;
+        return true;
+    }
+    
+    // Mutation activation was not canceled in legacy mutation effects, so return true.
+    return true;
+}
+
+void Character::deactivate_mutation(const muttype_id &flag)
+{
+    if (!has_mut(flag)) {
+        // We don't have the desired mutation!
+        debugmsg("Tried to deactivate a nonexistent mutation!");
+        return;
+    }
+    if (!my_mutations[flag].is_activatable()) {
+        // It's not an activatable mutation!
+        debugmsg("Tried to deactivate a non-activatable mutation!");
+        return;
+    }
+    if (!my_mutations[flag].is_active()) {
+        // It's already deactivated, abort
         return;
     }
+    if (is_player()) {
+        add_msg(m_neutral, _("You stop using your %s."), my_mutations[flag].get_name().c_str());
+    }
+    
+    // We've decided to deactivate here!
+    // Deactivate the mutation
+    my_mutations[flag].deactivate(*this)
+}
+
+void mutation::deactivate(Character &ch)
+{
+    muttype_id id = mut_type->id;
+    if (deactivation_effects(id, ch)) {
+        // We didn't stop the deactivation
+        charge = 0;
+    }
+}
+
+void mutation::deactivation_effects(const muttype_id id, Character &ch)
+{
+    /* NEW ON-DEACTIVATION MUTATION EFFECTS SHOULD GO HERE.
+     * Cast ch to player if you really need to access player class variables. */
 }
 
-void player::deactivate_mutation(std::string mut)
+void Character::process_mutations()
 {
-    traits[mut].powered = false;
+    // Bitset is <hunger, thirst, fatigue>
+    std::bitset<3> stops;
+    if (hunger >= 2800) {
+        stops.set(1);
+    }
+    if (thirst >= 520) {
+        stops.set(2);
+    }
+    if (fatigue >= 575) {
+        stops.set(3);
+    }
+    std::unordered_map<std::string, int> pro_costs;
+    for (auto &mut : my_mutations) {
+        if (!mut.second.process(pro_costs, stops) && is_player()) {
+            add_msg(m_info, _("You can't keep your %s active anymore!"), mut.get_name().c_str());
+        }
+    }
+    // Pay any reactivation costs
+    hunger += pro_costs["hunger"];
+    thirst += pro_costs["thirst"];
+    fatigue += pro_costs["fatigue"];
```

Comment:
Out of curiosity: why a `std::unordered_map`, why not a simple struct:
```C++
struct cost_t {
    int hunger = 0;
    int thirst = 0;
    int fatigue = 0;
};
```
A struct allows the compiler to check the names of the members, the map allows simple spelling errors like `costs["hunger "] += ...` (note the space, it's a different entry). And you never use a variable to access the map either, all keys used with that map are static strings.